src/hotspot/share/code/codeCache.cpp

```diff
@@ -678,7 +678,7 @@ void CodeCache::nmethods_do(void f(nmethod* nm)) {
 
 void CodeCache::metadata_do(MetadataClosure* f) {
   assert_locked_or_safepoint(CodeCache_lock);
-  NMethodIterator iter(NMethodIterator::only_alive_and_not_unloading);
+  NMethodIterator iter(NMethodIterator::only_alive);
```

Comment:
The `CodeCache::metadata_do` is used in `MetadataOnStackMark::MetadataOnStackMark`.
Besides `RedefineClasses` the `MetadataOnStackMark` is also used in `ClassLoaderDataGraph::walk_metadata_and_clean_metaspaces`.
Is the change at L681 correct for `ClassLoaderDataGraph::walk_metadata_and_clean_metaspaces`as well?

Reply:
Yes, in ClassLoaderDataGraph::walk_metadata_and_clean_metaspaces, it calls the version of MetadataOnStackMark that doesn't call CodeCache::metadata_do - it only walks the metadata that was saved from the full redefinition walk:

caller:
  MetadataOnStackMark md_on_stack(walk_all_metadata, /*redefinition_walk*/false);

callee:
    if (redefinition_walk) {
      // We have to walk the whole code cache during redefinition.
      CodeCache::metadata_do(&md_on_stack);
    } else {
      CodeCache::old_nmethods_do(&md_on_stack);
    }

During redefinition, we have to walk the unloaded nmethods as well as the alive nmethods.

Reply:
The nmethods in the table are removed when the unloaded nmethods are flushed so that the table doesn't contain stale entries.